src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -484,21 +480,27 @@ public OpenIdConnectAuthenticationHandler(HttpClient backchannel)
                 return null;
             }
 
+            message = authorizationCodeRedeemedContext.ProtocolMessage;
+
             // no need to validate signature when token is received using "code flow" as per spec [http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation].
             var validationParameters = Options.TokenValidationParameters.Clone();
             validationParameters.ValidateSignature = false;
 
-            ticket = ValidateToken(idToken, message, properties, validationParameters, out jwt);
+            ticket = ValidateToken(message.IdToken, message, properties, validationParameters, out jwt);
 
             await ValidateOpenIdConnectProtocolAsync(null, message);
 
             if (Options.GetClaimsFromUserInfoEndpoint)
             {
                 Logger.LogDebug(Resources.OIDCH_0040_Sending_Request_UIEndpoint);
-                ticket = await GetUserInformationAsync(properties, tokenEndpointResponse.Message, ticket);
+                ticket = await GetUserInformationAsync(tokenEndpointResponse.ProtocolMessage, ticket);
+                if (ticket == null)
+                {
+                    return null;
+                }
             }
 
-            var securityTokenValidatedContext = await RunSecurityTokenValidatedEventAsync(message, ticket);
+            var securityTokenValidatedContext = await RunIdTokenValidatedEventAsync(message, ticket, tokenEndpointResponse);
```

Comment:
Hum, `IdTokenValidatedContext.ProtocolMessage` will be the authorization response (basically, only containing the code and the state in this scenario). It's definitely not obvious and not really useful. Consider renaming this property to `AuthorizationResponse`.